feat(compiler-cli): add support to extend `angularCompilerOptions`

alan-agius4 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
`TypeScript` only supports merging and extending of `compilerOptions`. 

Issue Number: #22684

## What is the new behavior?
This is an implementation to support extending and inheriting of `angularCompilerOptions` from multiple files.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alan-agius4 commented:
Might need a bit of help here with `Bazel` I am noticing an error, however I cannot seem to find the root cause.

What I noticed is that in `Bazel` is seems the the extends paths are point to files which don't exists.

Examaple
```ts
'tsconfig' : 'bazel-out/darwin-fastbuild/bin/packages/platform-browser/_platform-browser.bundle_index.tsconfig.json',
'config.extends': '../../../../../packages/tsconfig-build'
'resolved extend': '/private/var/tmp/_bazel_alan/58002f7b65e603e10c8e51af4211aa59/bazel-sandbox/7299424352732944727/execroot/angular/packages/tsconfig-build',
 'error': 
   { file: undefined,
     start: undefined,
     length: undefined,
     messageText: 'The specified path does not exist: \'/private/var/tmp/_bazel_alan/58002f7b65e603e10c8e51af4211aa59/bazel-sandbox/7299424352732944727/execroot/angular/packages/tsconfig-build.json\'.',
     category: 1,
     code: 5058 } }
```

alan-agius4 commented on packages/compiler-cli/test/perform_compile.spec.ts:
```diff
@@ -0,0 +1,50 @@
+/**
```
I am almost certain that these tests are never running :(

alan-agius4 commented on packages/compiler-cli/test/BUILD.bazel:
```diff
@@ -132,3 +132,30 @@ jasmine_node_test(
 )
+
+# perform_compile_spec
+ts_library(
```
not quite sure, why this is per every spec file. Rather than having a single bazel setup that globsall spec files.

alexeagle replied:
I think it's like this to manually shard the targets for faster execution since they can run in parallel. It does seem like overkill to me

alan-agius4 commented:
//cc @chuckjaz

alan-agius4 commented:
Bump

alan-agius4 commented:
Bump @chuckjaz & @mhevery

alan-agius4 commented:
bump

alan-agius4 commented:
Any feedback please?

alan-agius4 commented:
Bump @vicb and @alxhub

alxhub commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -0,0 +1,58 @@
+    basePath = support.basePath;
+  });
+
+  function writeSomeRoutes() {
```
`writeSomeConfigs`?

alan-agius4 commented:
Travis error looks to be a flake.

IgorMinar reviewed (dismissed):
Can we please add a documentation for this change? we shouldn't merge new features like this without docs. thanks!

the current compiler options are documented here: aio/content/guide/aot-compiler.md

alan-agius4 commented:
Sure, though in all fairness I am sure sure if it fits under compiler options, but I can add this it there non the less.

alan-agius4 commented:
@IgorMinar, I have added the requested docs. Can you please review them?

Please feel free to suggest or do any changes.

Ps: I have also rebased.

benlesh commented:
@alan-agius4 or @IgorMinar ... flag this as `PR action: merge` when you're done with your review process, please. I'll keep an eye out.

benlesh commented:
For googlers: [presubmit](http://test/OCL:208738053:BASE:208738078:1534290176381:a45f5eda)

alan-agius4 commented:
@IgorMinar, do you mind re-checking this please? Thanks.

alexeagle commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -128,7 +128,37 @@ export function readConfiguration(
 
-    let {config, error} = ts.readConfigFile(projectFile, ts.sys.readFile);
+    const readExtendedConfigFile =
+        (configFile: string, existingConfig?: any): {config?: any; error?: ts.Diagnostic;} => {
```
I think we prefer to comma-separate

alan-agius4 replied:
updated

alexeagle commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -128,7 +128,37 @@ export function readConfiguration(
+
+            if (fs.existsSync(extendedConfigPath)) {
+              // Call read config recursively as TypeScript only merges CompilerOptions
+              return readExtendedConfigFile(extendedConfigPath, baseConfig);
```
yum, we are missing the recursive case in tsc_wrapped https://github.com/bazelbuild/rules_typescript/blob/d4873377218c49d382275546d11fe590f2fa2ef0/internal/tsc_wrapped/tsconfig.ts#L275-L300

mary-poppins commented:
You can preview 42960a0 at https://pr22717-42960a0.ngbuilds.io/.
You can preview b691788 at https://pr22717-b691788.ngbuilds.io/.
You can preview 6881ca4 at https://pr22717-6881ca4.ngbuilds.io/.
You can preview 67495a8 at https://pr22717-67495a8.ngbuilds.io/.
You can preview 60d7910 at https://pr22717-60d7910.ngbuilds.io/.
You can preview 9155878 at https://pr22717-9155878.ngbuilds.io/.
You can preview 3edd13c at https://pr22717-3edd13c.ngbuilds.io/.
You can preview bd4e08d at https://pr22717-bd4e08d.ngbuilds.io/.
You can preview 3a9a286 at https://pr22717-3a9a286.ngbuilds.io/.
You can preview 3cc93f5 at https://pr22717-3cc93f5.ngbuilds.io/.
You can preview 6f1d0ef at https://pr22717-6f1d0ef.ngbuilds.io/.
You can preview 2d154d2 at https://pr22717-2d154d2.ngbuilds.io/.
You can preview 1ad54f2 at https://pr22717-1ad54f2.ngbuilds.io/.
You can preview 99682dc at https://pr22717-99682dc.ngbuilds.io/.
You can preview fa5ff53 at https://pr22717-fa5ff53.ngbuilds.io/.
You can preview 89ca5ff at https://pr22717-89ca5ff.ngbuilds.io/.
You can preview 41cc797 at https://pr22717-41cc797.ngbuilds.io/.
You can preview 73e475c at https://pr22717-73e475c.ngbuilds.io/.

jenniferfell commented:
@alan-agius4 Thanks. We've been working on a related doc PR that is also about ready to merge. It contains some improvements to the order of sections within the AOT doc: https://github.com/angular/angular/pull/22353

I think things will merge more seamlessly if we land 22353 first. I've ask the author to accelerate and she should be done with final changes by EOD today. 

Meanwhile, I'll take a look at your new section with an eye toward where it will fit best in the new structure and do a quick copy edit. 

Will that work for you?

alan-agius4 commented:
Yes, that will work.

On Fri, 07 Sep 2018 at 21:02, Jennifer Fell <notifications@github.com>
wrote:

> @alan-agius4 <https://github.com/alan-agius4> Thanks. We've been working
> on a related doc PR that is also about ready to merge. It contains some
> improvements to the order of sections within the AOT doc: #22353
> <https://github.com/angular/angular/pull/22353>
>
> I think things will merge more seamlessly if we land 22353 first. I've ask
> the author to accelerate and she should be done with final changes by EOD
> today.
>
> Meanwhile, I'll take a look at your new section with an eye toward where
> it will fit best in the new structure and do a quick copy edit.
>
> Will that work for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it orn GitHub
> <https://github.com/angular/angular/pull/22717#issuecomment-419533643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQv-Wkt5OEdPQ8bIg_eYHWlZKUsrvowvks5uYsKxgaJpZM4SnJm5>
> .
>

mary-poppins commented:
You can preview ab8f729 at https://pr22717-ab8f729.ngbuilds.io/.

mary-poppins commented:
You can preview 70a5519 at https://pr22717-70a5519.ngbuilds.io/.

mary-poppins commented:
You can preview 3f594fa at https://pr22717-3f594fa.ngbuilds.io/.

mary-poppins commented:
You can preview 4f79913 at https://pr22717-4f79913.ngbuilds.io/.

alan-agius4 committed:
feat(compiler-cli): add support to extend `angularCompilerOptions`

`TypeScript` only supports merging and extending of `compilerOptions`. This is an implementation to support extending and inheriting of `angularCompilerOptions` from multiple files.

Closes: #22684

mary-poppins commented:
You can preview 6c5ce1f at https://pr22717-6c5ce1f.ngbuilds.io/.

alan-agius4 commented:
@jenniferfell CI is all green here as well

IgorMinar reviewed (approved):
thanks for updating the pr

IgorMinar commented:
http://test/OCL:213668215:BASE:213668331:1537383459910:513bda2c
